synapse/server_notices/server_notices_manager.py

```diff
@@ -159,3 +149,55 @@ def get_notice_room_for_user(self, user_id):
 
         logger.info("Created server notices room %s for %s", room_id, user_id)
         return room_id
+
+    @defer.inlineCallbacks
+    def _get_existing_notice_room_for_user(self, user_id):
+        # type: (str) -> defer.Deferred[Optional[str]]
+        """Retrieve the ID for the existing notice room for the given user ID if it
+        exists.
+
+        Returns:
+            The ID of the notice room, None if no room exist for this user.
+        """
+        room_id = None
+
+        rooms = yield self._store.get_rooms_for_local_user_where_membership_is(
+            user_id, [Membership.INVITE, Membership.JOIN]
+        )
+        for room in rooms:
+            # it's worth noting that there is an asymmetry here in that we
+            # expect the user to be invited or joined, but the system user must
+            # be joined. This is kinda deliberate, in that if somebody somehow
+            # manages to invite the system user to a room, that doesn't make it
+            # the server notices room.
+            user_ids = yield self._store.get_users_in_room(room.room_id)
+            if self.server_notices_mxid in user_ids:
+                # we found a room which our user shares with the system notice
+                # user
+                logger.info("Using room %s", room.room_id)
+                room_id = room.room_id
+
+        return room_id
+
+    @defer.inlineCallbacks
+    def invite_user_in_notice_room(self, user_id: str, room_id: str):
+        """Invite the given user to the given server notices room.
+
+        Args:
+            user_id: The ID of the user to invite.
+            room_id: The ID of the room to invite the user to.
+        """
+        requester = create_requester(self.server_notices_mxid)
+
+        # Make sure a notice room doesn't already exist for this user, otherwise the auth
+        # checks will make the current request (i.e. sync) fail.
+        existing_room_id = yield self._get_existing_notice_room_for_user(user_id)
+        if existing_room_id:
+            return
```

Comment:
I'm really struggling to get my head around this logic. We pass in a room id, which we know is a server notices room, but if it turns out that a *different* room is a server notices room, we don't do anything? And this behaviour isn't described in the docstring?

Reply:
No, the issue here is that if the user has already joined or been invited to the room, inviting them will fail (and therefore make the sync request or whatever is triggering this function to be called return an auth error because of failing auth checks). I'll rephrase the comment to make it clearer, and mention that behaviour in the docstring.

The reason we need to do this "create vs invite" split is that, as I mention in the initial comment of this PR, the MAU alerting code calls `get_or_create_notice_room_for_user` before knowing whether it should send a notice, and does so on each sync (to check if a) the limit has been reached and b) the user has already been notified). Therefore, if this function invites the user, users that aren't already in a room with the server notices user will be invited to one at each restart of the server (since that function has a cache which entries are never invalidated, so it's only cleared when the server restarts).

Another bug this split fixes is that, since the result of that function is cached, if a user leaves the room and the server needs to send them a notice, it would do so without ensuring the room it's about to send the notice to has been joined by the user, or at least that the user has been invited to it.

I agree that it would be better if the MAU alerting code didn't call `get_or_create_notice_room_for_user` without knowing that it's going to notify the user, and if we could drop the cache on `get_or_create_notice_room_for_user`, but I'm not sure we can do either (or at least not in a non-trivial way).

Reply:
Done as part of https://github.com/matrix-org/synapse/pull/7199/commits/3da9f9a65367b98854052bd2f4cc752bb6bbd826